ql/src/java/org/apache/hadoop/hive/ql/exec/repl/bootstrap/load/table/LoadPartitions.java

```diff
@@ -224,14 +225,26 @@ private void addPartition(boolean hasMorePartitions, AddPartitionDesc addPartiti
             + partSpecToString(partSpec.getPartSpec()) + " with source location: "
             + partSpec.getLocation());
 
-    Path tmpPath = PathUtils.getExternalTmpPath(replicaWarehousePartitionLocation, context.pathInfo);
+    Path tmpPath = replicaWarehousePartitionLocation;
+
+    // if move optimization is enabled, copy the files directly to the target path. No need to create the staging dir.
+    LoadFileType loadFileType;
+    if (event.replicationSpec().isInReplicationScope() &&
+            context.hiveConf.getBoolVar(REPL_ENABLE_MOVE_OPTIMIZATION)) {
+      loadFileType = LoadFileType.IGNORE;
+    } else {
+      loadFileType =
+              event.replicationSpec().isReplace() ? LoadFileType.REPLACE_ALL : LoadFileType.OVERWRITE_EXISTING;
+      tmpPath = PathUtils.getExternalTmpPath(replicaWarehousePartitionLocation, context.pathInfo);
+    }
+
     Task<?> copyTask = ReplCopyTask.getLoadCopyTask(
         event.replicationSpec(),
         sourceWarehousePartitionLocation,
         tmpPath,
-        context.hiveConf
+        context.hiveConf, false, false
```

Comment:
needCmRecycle flag cannot be false always. For insert overwrite case, need to recycle it.

Reply:
do we need to do cm recycle during boot strap ?